src/community/web-resources/src/main/java/org/geoserver/web/treeview/TreeView.java

```diff
@@ -0,0 +1,417 @@
+            return getNearestViewInternal(selectedNodeModel.getObject());
+        } else {
+            return null;
+        }
+    }
+    
+    /**
+     * Get the view for a node, if there is not yet a view for this node, it will return the nearest parent node.
+     * Refreshing this view refreshes all child nodes as well.
+     * It is efficient to only refresh the necessary node(s) on ajax requests, rather than the whole tree. 
+     */
+    public Panel getNearestView(TreeNode<T> node) {
+        return getNearestViewInternal(node);
+    }
+    
+    /**
+     * Add a selection listener
+     */
+    public void addSelectionListener(SelectionListener<T> listener) {
+        selectionListeners.add(listener);
+    }
+    
+    /**
+     * Change selected node, expand if necessary. Events aren't called. Caller is responsible for refreshing the view(s).
+     * 
+     */
+    public void setSelectedNode(TreeNode<T> selectedNode) {
+        setSelectedNode(selectedNode, null);
+    }
+    
+    /**
+     * Change selected node, expand if necessary, fire events, and refresh the whole treeview.
+     * 
+     */
+    public void setSelectedNode(TreeNode<T> selectedNode, AjaxRequestTarget target) {
+        //expand if necessary
+        if (selectedNode != null) {
+            TreeNode<T> node = selectedNode.getParent();
+            while (node != null) {
+                node.getExpanded().setObject(true);
+                node = node.getParent();     
+            }
+        }
+        setSelectedNodeInternal(selectedNode, target);
+        if (target != null) {
+            target.add(this);
+        }
+    }
+    
+    /**
+     * Register a customised mark. The style of the mark can be defined in CSS with ".css-treeview a.{markName}"
+     * 
+     * @param markName name of the mark
+     */
+    public void registerMark(String markName) {
+        marks.put(markName, new Mark(markName));
+    }
+    
+    /**
+     * Add marked node to a mark.
+     * 
+     * @param markName name of the mark
+     * @param node node to be added
+     */
+    public void addMarked(String markName, TreeNode<T> node) {
+        marks.get(markName).getMarked().add(node.getUniqueId());
+    }
+    
+    /**
+     * Clear all marked nodes from a mark.
+     * 
+     * @param markName name of the mark to be cleared
+     */
+    public void clearMarked(String markName) {
+        marks.get(markName).getMarked().clear();
+    }
+    
+    /**
+     * Check if a node has a mark
+     * 
+     * @param markName name of the mark
+     * @param node the node that may or may not be marked
+     * @return whether the mode is marked
+     */
+    public boolean hasMark(String markName, TreeNode<T> node) {
+        return marks.get(markName).getMarked().contains(node.getUniqueId());
+    }
+    
+    //------------------- internal methods/classes    
+        
+    protected void setSelectedNodeInternal(TreeNode<T> selectedNode, AjaxRequestTarget target) {
```

Comment:
If these are really internal, should they be private? If not, then perhaps some javadoc is in order to aid reuse by subclasses.